### PR TITLE
raftstore: Add apply snapshot observer 

### PR DESCRIPTION
**Pull request by MyonKeminta**
###  What have you changed?

This PR adds a new observer named `ApplySnapshotObserver`, which is triggered when apply snapshot happens. When applying from plain file, `pre_apply_plain_key` will be invoked for each key it writes to DB. When applying from sst file, `pre_apply_sst` will be invoked for each file.

This will be used for Green GC to monitor apply snapshot events on Lock CF. Lock CF are always transfered via plain files, so we get those keys while applying snapshot via `pre_apply_plain_key`.

- [x] Implementation
- [x] Tests

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: src/raftstore/store/snap.rs [hunk 1]
```diff
@@ -852,10 +856,21 @@ impl Snapshot for Snap {
             if plain_file_used(cf_file.cf) {
                 let path = cf_file.path.to_str().unwrap();
                 let batch_size = options.write_batch_size;
-                snap_io::apply_plain_cf_file(path, &abort_checker, &options.db, cf, batch_size)?;
+                // let cb = |key| {
```
Comment (nrc) [inline src/raftstore/store/snap.rs]: please remove commented out code

File: src/raftstore/store/snap/io.rs [hunk 2]
```diff
@@ -98,13 +98,17 @@ pub fn build_sst_cf_file(
 }
 
 /// Apply the given snapshot file into a column family.
```
Comment (nrc) [inline src/raftstore/store/snap/io.rs]: Can you document when `key_callback` is called and what is passed to it please

File: src/raftstore/store/snap.rs (hunk 1 above)
Comment (MyonKeminta) [reply src/raftstore/store/snap.rs]: Sorry. This is a mistake. I'll update this PR later and the implementation will be slightly changed.

**Comment by MyonKeminta** [comment]
@nrc Thanks for your review, but I found there's still some problem in it. I'll mark it DNM for now.

File: src/raftstore/coprocessor/dispatcher.rs [hunk 3]
```diff
@@ -215,6 +226,30 @@ impl CoprocessorHost {
         }
     }
 
+    pub fn pre_apply_plain_keys_from_snapshot(
+        &self,
+        region: &Region,
+        cf: CfName,
+        key: &[(Vec<u8>, Vec<u8>)],
```
Comment (BusyJay) [inline src/raftstore/coprocessor/dispatcher.rs]: There are not just keys.

File: src/raftstore/coprocessor/dispatcher.rs [hunk 4]
```diff
@@ -350,6 +385,23 @@ mod tests {
...
+            &self,
+            ctx: &mut ObserverContext<'_>,
+            _: CfName,
+            _: &[(Vec<u8>, Vec<u8>)],
+        ) {
+            self.called.fetch_add(9, Ordering::SeqCst);
+            ctx.bypass = self.bypass.load(Ordering::SeqCst);
+        }
+
+        fn pre_apply_sst(&self, ctx: &mut ObserverContext<'_>, _: CfName) {
```
Comment (BusyJay) [inline src/raftstore/coprocessor/dispatcher.rs]: Are SST file paths necessary?

Comment (MyonKeminta) [reply src/raftstore/coprocessor/dispatcher.rs]: Good idea

**Comment by MyonKeminta** [comment]
@BusyJay PTAL again, thx

**Pull request merged**